### PR TITLE
EVG-1966 add event logging for degraded mode

### PR DESCRIPTION
**Pull request by john-m-liu**
There will be a wrapper in the routes (coming next) that sets the data in the admin document and also logs an event. I wanted to put it in the model package, but that would create an import loop

### CODE REVIEW

File: model/admin/db.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 
 const (
 	Collection          = "admin"
-	systemSettingsDocID = "global"
+	SystemSettingsDocID = "global"
```
Comment (tychoish) [inline model/admin/db.go]: what purpose is there in exporting this?

File: model/admin/db.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ var (
 	schedulerKey     = bsonutil.MustHaveTag(ServiceFlags{}, "SchedulerDisabled")
 )
 
-var settingsQuery = db.Query(bson.M{idKey: systemSettingsDocID})
+var settingsQuery = db.Query(bson.M{idKey: SystemSettingsDocID})
```
Comment (tychoish) [inline model/admin/db.go]: I guess there's no real problem with this but I find the queries-in-variables approach to be weird.

File: model/event/event_finder.go [hunk 3]
```diff
@@ -111,6 +111,14 @@ func RecentSchedulerEvents(distroId string, n int) db.Q {
 	return SchedulerEventsForId(distroId).Sort([]string{"-" + TimestampKey}).Limit(n)
 }
 
+// Admin Events
+// RecentAdminEvents returns the N most recent admin events
+func RecentAdminEvents(n int) db.Q {
+	return db.Query(bson.D{
+		{DataKey + "." + ResourceTypeKey, ResourceTypeAdmin},
+	}).Sort([]string{"-" + TimestampKey}).Limit(n)
```
Comment (tychoish) [inline model/event/event_finder.go]: There's no real reason to use a bson.D here, and a bson.M might make more sense?

File: model/admin/db.go (hunk 1 above)
Comment (john-m-liu) [inline model/admin/db.go]: This is used for the resource ID field when logging an event

File: model/event/event_finder.go (hunk 3 above)
Comment (john-m-liu) [inline model/event/event_finder.go]: Fixed

File: model/admin/db.go (hunk 1 above)
Comment (tychoish) [inline model/admin/db.go]: I would try to keep these concepts separate, the value in model/admin is super important (though we could change it at some point,) and if we changed it, we probably wouldn't want to change the value in the event log?

Comment (john-m-liu) [inline model/admin/db.go]: changed it back to private

File: model/event/admin_event.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+func (evt AdminEventData) IsValid() bool {
+	return evt.ResourceType == ResourceTypeAdmin
+}
+
+// logAdminEventBase is a helper function to log an admin event
+func logAdminEventBase(eventType string, eventData AdminEventData) error {
+	eventData.ResourceType = ResourceTypeAdmin
+	event := Event{
+		Timestamp:  time.Now(),
+		ResourceId: "",
```
Comment (tychoish) [inline model/event/admin_event.go]: nit, no need to specify the default value.

File: model/event/admin_event.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+	logger := NewDBEventLogger(AllLogCollection)
+	if err := logger.LogEvent(event); err != nil {
+		return errors.Wrap(err, "Error logging admin event")
+	}
+	return nil
+}
+
+// LogBannerChanged will log a change to the banner field
+func LogBannerChanged(oldText, newText string) error {
```
Comment (bsamek) [inline model/event/admin_event.go]: Can this be private?

File: model/event/admin_event.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	if oldText == newText {
+		return nil
+	}
+	return logAdminEventBase(BannerChanged, AdminEventData{OldVal: oldText, NewVal: newText})
+}
+
+// LogServiceChanged will log a change to the service flags
+func LogServiceChanged(oldFlags, newFlags admin.ServiceFlags) error {
```
Comment (bsamek) [inline model/event/admin_event.go]: Can this be private?

File: model/event/event_finder.go [hunk 7]
```diff
@@ -111,6 +111,14 @@ func RecentSchedulerEvents(distroId string, n int) db.Q {
 	return SchedulerEventsForId(distroId).Sort([]string{"-" + TimestampKey}).Limit(n)
 }
 
+// Admin Events
+// RecentAdminEvents returns the N most recent admin events
+func RecentAdminEvents(n int) db.Q {
```
Comment (bsamek) [inline model/event/event_finder.go]: Can this be private?

File: model/event/admin_event.go (hunk 5 above)
Comment (john-m-liu) [inline model/event/admin_event.go]: This will be called by the route

File: model/event/admin_event.go (hunk 6 above)
Comment (john-m-liu) [inline model/event/admin_event.go]: This will be called by the route

File: model/event/event_finder.go (hunk 7 above)
Comment (john-m-liu) [inline model/event/event_finder.go]: Yep, I'll change this in master directly

Comment (bsamek) [inline model/event/event_finder.go]: Thanks, sgtm.

**Pull request merged**